### DISCUSSION TITLE
Add MediaPicker to .NET MAUI Community Toolkit

### DISCUSSION THREAD

**Discussion by jfversluis** [CommunityToolkit/Maui#233] [New Feature Discussions]
Right now the media picker functionality is a bit all over the place. There is the original [MediaPlugin](https://github.com/jamesmontemagno/MediaPlugin) by James, but as far as I know that is no longer maintained since no one stepped up.

That evolved into it being added to Xamarin.Essentials, but only the basic functionality made it over and with Essentials being in maintenance mode, we don't have the possibility to add anything to that anymore.

This all resulted into the great efforts of @dimonovdd who has created his [Xamarin.MediaGallery](https://github.com/dimonovdd/Xamarin.MediaGallery) plugin.

Personally I think this should be part of code .NET MAUI (Essentials) but the truth is we're not there for the upcoming GA version, so also that is sometime into the future. 

In an effort to make this available to people as fast as possible while also having an easy transition path into core .NET MAUI, my proposal would be get proper media picker functionalities in the .NET MAUI Community Toolkit. That way we will have it around the same time that .NET MAUI lands and hopefully we will be able to transition it soon and easy to .NET MAUI itself.

I have already spoken to @dimonovdd a bit as he offered some help earlier. I hope we can still count on him and/or others to pitch in with this one. What are peoples thoughts? And what is stuff that should at least be in there? What is your wishlist from high priority to low?

**Comment by dimonovdd** [comment]
@jfversluis Hi. I was planning to move [MediaGallery](https://github.com/dimonovdd/Xamarin.MediaGallery) to `MAUI.Essentials`, but I still didn't understand [the roadmap of Essentials](https://github.com/dotnet/maui/discussions/2910), so I added MAUI support for [MediaGallery](https://github.com/dimonovdd/Xamarin.MediaGallery).

[As I said before,](https://github.com/CommunityToolkit/Maui/discussions/33) I don't see the point in moving `MediaPicker` or `Contacts`, `FilePicker`,  `WebAuthenticator` to `Community Toolkit`. Perhaps we could make a separate project for these features Or do nothing until we know the roadmap of Essentials.

**Reply by jfversluis** [reply]
OK, lets make a few things clear. 

> I don't see the point in moving MediaPicker or Contacts, FilePicker, WebAuthenticator to Community Toolkit

This is not happening anymore as far as I know. What is in Xamarin Essentials right now will stay there and will be moved to .NET MAUI Essentials. In fact, I think it already is.

However, for MediaPicker I see there is a lot of attention for it and rightfully so. People were used to [MediaPlugin](https://github.com/jamesmontemagno/MediaPlugin) and are now expecting (again, rightfully so) that all the features would be in Xamarin Essentials. Unfortunately, only the very core was implemented and now people would love to see the rest implemented in Essentials as well.

The problem is that we can't invest in Xamarin Essentials anymore at this time. We can only add bugfixes, just like Xamarin.Forms. Also, what we have committed to building for .NET MAUI Essentials is already determined. What we have committed to build is porting over Xamarin Essentials to .NET MAUI Essentials and that will be it for GA. That means, we can't add any of the new MediaPicker stuff to .NET MAUI Essentials at this time.

That in its turn means that the _earliest_ we can add the much sought after MediaPicker functionalities (rotation, control over quality, etc) is in .NET 7, which will be in November of this year. And to be honest with the backlog we have now, I'm not sure if we will make that and then it will be .NET 8 in November 2023.

Your MediaGallery is awesome. If that has .NET MAUI support, even more awesome. That means we can point people to that. And if that is what you want to continue to do, that is fine. The only thing I was wondering is if we can maybe make your MediaGallery part of the .NET MAUI Community Toolkit.

The Toolkit can be released independently of .NET MAUI (Essentials). If we start adding the MediaGallery to .NET MAUI Community Toolkit right now, we can even release it by the time .NET MAUI is released. I know a lot of people would love less dependencies and a package that is somehow related to Microsoft. That is part of why we started with the Community Toolkit in the first place. By making MediaGallery part of .NET MAUI Community Toolkit people will have one less package to install and there is governance by the .NET Foundation and people from Microsoft are working on this.

Another advantage would be that the step from .NET MAUI Community Toolkit to .NET MAUI (Essentials) is a bit easier than from a third-party package. And you have already expressed your wish of making this code part of .NET MAUI so I thought this would make most sense to do. Because I totally agree with that, this should be part of .NET MAUI!

What I'm trying to do here is find a way so we can get this to people as fast as possible while also considering a path forward for the future of this functionality. It goes without saying that you can keep being involved as the Community Toolkit is open to anyone. But that's all up to you. Just putting this discussion out there to hear everyones thoughts.

Hope this makes it a bit more clear :)

**Reply by dimonovdd** [reply]
@jfversluis I don't understand why it should be transferred to Toolkit? It will be the 4th implementation. 
I was making my plugin as a temporary solution until MAUI is released. 
I can give the rights to the project or the nuget package to other people, Microsoft or .NET Foundation if you have a desire

**Reply by dimonovdd** [reply]
1) [MediaPlugin](https://github.com/jamesmontemagno/MediaPlugin) by @jamesmontemagno
2) Xamarin Essentials
3) MAUI Essentials
4) [Xamarin.MediaGallery](https://github.com/dimonovdd/Xamarin.MediaGallery) is clone from Essentials with [SaveMedia](https://github.com/xamarin/Essentials/pull/1628) and Bug Fixes
5) Community Toolkit is clone Xamarin.MediaGallery

And if we add something or fix some bug in one project, we will have to make these changes to `MediaGallery`, `Community Toolkit` and `MAUI Essentials`. 

I think we can wait for the release of `MAUI Essentials` and just transfer `MediaGallery` to it

**Reply by pictos** [reply]
@dimonovdd if we bring the `Xamarin.MediaGallery` to `Community Toolkit (MCT)` why we still need to update and support the `Xamarin.MediaGallery`? 
Also, if we have it here on MCT we can talk with the MAUI team to say that MCT will be the way to go if you want that feature. When this code be mature and stable enough we will port it to MAUI and deprecate it here.

So at the end of the day we will have to maintain the `Xamarin.Essentials` (just for bug fixes) and `MCT`. Or am I missing something?

**Reply by dimonovdd** [reply]
> So at the end of the day we will have to maintain the `Xamarin.Essentials` (just for bug fixes) and `MCT`. Or am I missing something?

Fixes require serious improvements and it are not accepted by Microsoft Team

>@dimonovdd if we bring the Xamarin.MediaGallery to Community Toolkit (MCT) why we still need to update and support the Xamarin.MediaGallery?

Some developers have recently switched from Essentials to MediaGallery. And they will need to switch to a new implementation  again.

**Reply by pictos** [reply]
> Some developers have recently switched from Essentials to MediaGallery. And they will need to switch to a new implementation again.


Well, we're talking about Maui, not Forms. When devs migrate from Forms > MAUI they will need to migrate to another package and handle some breaking changes. So I can't see this as an issue.

> Fixes require serious improvements and it are not accepted by Microsoft Team

I can't talk too much about it because I don't have all the context. From my point of view the MS team is very open to bug fixes that will not introduce breaking changes or will add too much complexity.

**Reply by dimonovdd** [reply]
> I can't talk too much about it because I don't have all the context. From my point of view the MS team is very open to bug fixes that will not introduce breaking changes or will add too much complexity.

You can see how many Pull Request to Essentials are open. Some of them are over a year old

> Well, we're talking about Maui, not Forms.

 Xamarin.MediaGallery support Xamarin (Forms and Native) and MAUI and I need to support it

**Reply by jfversluis** [reply]
> You can see how many Pull Request to Essentials are open. Some of them are over a year old

You can keep bringing this up, that won't change anything. We're working on it with the capacity we have and for doing better with .NET MAUI. Also, I don't feel that is relevant to this discussion. There will be bugfixes for Xamarin.Forms and Xamarin Essentials. 

However, the MediaPicker stuff isn't a bug, it's just functionality that is not implemented. And one thing we won't do (for Forms and Essentials) is add new functionality because everything we add we also need to port to .NET MAUI and maintain on both sides eating into that same capacity.

My proposal is as simple as:
1. Move the (.NET MAUI) code of MediaGallery into the .NET MAUI Community Toolkit
2. Stop the MediaGallery for .NET MAUI, keep it for Xamarin.Forms as there is no real alternative from the Microsoft side right now
3. Maintain MediaGallery from the .NET MAUI Community Toolkit and from there transition it to .NET MAUI (Essentials) as soon as possible.

pros:
- It will be in the Toolkit which a lot of people are going to install already, so they need to only install the Toolkit and not also MediaGallery
- The Toolkit is backed by Microsoft which usually means a lot, especially to enterprises
- Maintenance (of MediaGallery) is not dependant on 1 person but is now shared with the .NET MAUI Community Toolkit team

cons: 
- People might've been using MediaGallery at this moment and need to switch to the Toolkit, although I would doubt that people are heavily using it for .NET MAUI already.

Let's turn this around. @dimonovdd what is your take on this? What would you like to do?

**Reply by dimonovdd** [reply]
My opinion may be wrong, we can create a survey and get the opinion of more developers, But I look at this problem from two sides:
1) Developer of real applications
2) Creator of MediaGallery

> Stop the MediaGallery for .NET MAUI, keep it for Xamarin.Forms as there is no real alternative from the Microsoft side right now

The code for Xamarin and MAUI has no differences.

1) As developer of real applications:
If I used MediaGallery in my Xamarin application I don't have to change anything when I switch from Xamarin to MAUI. Or will I have to add the new nugget (MAUI Community), figure out its subtleties and settings and conduct additional testing. Only 1 of my last 5 apps use Xamarin Community Toolkit, and 4 use MediaGallery.

2) As creator:
I have no need to support and monitor two implementations.

When I created this project, I wanted to simplify the life of developers. I have no desire to force them to choose from 5 implementations and perform additional actions when moving from Xamarin to MAUI.

MediaGallery looks like a great incubator to finish this features for MAUI Essentials for me. I would be very happy if this implementation was maintained by other developers from the community and Microsoft. I have no desire to retain the authorship, I just want these features to work well out of the box.

I really don't see the point in transferring the code of MediaGallery to .NET MAUI Community Toolkit. But I will do it if you provide real arguments.

**Comment by jfversluis** [comment]
> As developer of real applications:
> If I used MediaGallery in my Xamarin application I don't have to change anything when I switch from Xamarin to MAUI. Or will I have to add the new nugget (MAUI Community), figure out its subtleties and settings and conduct additional testing. Only 1 of my last 5 apps use Xamarin Community Toolkit, and 4 use MediaGallery.

This is indeed true for you, but if we look at numbers the Toolkit has way more downloads than MediaGallery (674.5k vs 18k), for Xamarin that is, but I expect the same for .NET MAUI as soon as it's released. This definitely is no definite proof that all people that use the Toolkit also have MediaGallery installed, but chances are that they already have the Toolkit and can now eliminate another dependency; MediaGallery.

And like I said earlier: I think you're overlooking the fact that developers working at enterprises feel much more comfortable with a library maintained by a team (and even Microsoft and the .NET Foundation) than a library that is maintained by just 1 person, in fact, some companies forbid the use of those libraries.

Don't get me wrong, I am in no way saying that your library is bad or how you're doing things is bad, I mean, the fact that I'm even considering asking you to help us with this is meant to be a great compliment because I love the work that you do! The only thing I want to achieve is getting this code to as many people as possible and into .NET MAUI Essentials as fast and easy as possible. And as you have expressed, that is also something you want.

> As creator:
> I have no need to support and monitor two implementations.

You still don't need to if we drop the MediaGallery .NET MAUI version and move that part to the Toolkit. Then you only have to maintain the Xamarin version and, if you still want, the .NET MAUI version but that only is on a different repository.

> MediaGallery looks like a great incubator to finish this features for MAUI Essentials for me.

No argument there, but I think the Toolkit is an even _better_ incubator. We already follow the same project structure as .NET MAUI so moving it will mostly be copy/paste. Developers using the MediaGallery or Toolkit then only need to change namespaces and not even that because of the net C# 10 templates we have, that should take care of that. And with the Toolkit we're closer to the people that are actually building .NET MAUI (Essentials). That means it's much easier in terms of moving it over and the timing of that.

But, and don't mistake this for being rude, I feel that if I haven't convinced you by now I won't be able to convince you, which is fine. Just to be clear, that might mean (note: might, we haven't made a decision yet) that we still might move forward with this, just not with your help.

In either case, I thank you for your input which is very insightful and of course; all the great work that you have done for the community and on the MediaGallery library!

**Reply by dimonovdd** [reply]
> but if we look at numbers the Toolkit has way more downloads than MediaGallery (674.5k vs 18k)

If we started looking at the numbers, then we should take into account that [Xamarin.MediaGallery.Permision.Maui](https://www.nuget.org/packages/Xamarin.MediaGallery.Permision.Maui/) has 5.3k downloads, and [CommunityToolkit.Maui](https://www.nuget.org/packages/CommunityToolkit.Maui/) only 1.3k

> I think you're overlooking the fact that developers working at enterprises feel much more comfortable with a library maintained by a team (and even Microsoft and the .NET Foundation) than a library that is maintained by just 1 person, in fact, some companies forbid the use of those libraries.

I know a project that was banned from using Xamarin.Essentials due to the large number of bugs and a very long time of fixing these bugs, but that project often used small libraries of single authors, such libraries are more flexible and quickly process bug fix requests.

> You still don't need to if we drop the MediaGallery .NET MAUI version and move that part to the Toolkit. Then you only have to maintain the Xamarin version and, if you still want, the .NET MAUI version but that only is on a different repository.

MediaGallery does not have a separate version for Xamarin or MAUI - it is the same code and the same nuget package, the difference is only 2 files, the main difference is only `Usings` presented in [this file](https://github.com/dimonovdd/Xamarin.MediaGallery/blob/main/MediaGallery/Usings/GlobalUsings.shared.cs).

You are also forgetting the fact that MAUI is not yet released and developers are currently using Xamarin and need these features in their Xamarin applications now and the ability to easily migrate from Xamarin to MAUI. Unfortunately, this was not possible before the release of MediaGallery.

Even if I or someone else adds these functions to the MAUI Toolkit - all this will have to be kept in two places, if something is fixed or changed in one project - then it will have to be changed in another. Why make it so unclear, whereas if we fix a bug or make improvements for MediaGallery, we immediately do it for both Xamarin and MAUI.  Moving the code this way will only add more confusion.

There is also the implementation of Xamarin and MAUI Essentials, which, to my great regret, are not finalized and have real bugs and problems. Developers of real applications simply will not have an understanding of which implementation to choose in order to get the most positive experience.

In my opinion, it would be most convenient to continue supporting MediaGallery and transfer it directly to MAUI Essentials, I can do this with great pleasure and try to do it as efficiently as possible. I really don't understand the point of adding this to the MAUI Toolkit, it looks like you just want to transfer it under Microsoft's control, since it's easier for Microsoft, and not to make it better by fixing and implementing these functions immediately in MAUI Essentials.

I could give the full access to the repository and nuget package to you personally or to Microsoft.

And we have even closed one [discussion](https://github.com/CommunityToolkit/Maui/discussions/33) and [issue](https://github.com/CommunityToolkit/Maui/issues/11) related to adding MediaPicker to this Toolkit.  But what you proposed looks strange for me, and it seems to me that these ideas are based on the lack of resources at Microsoft for Xamarin support and MAUI development.


As I said when I was doing this project, I had hoped for a long time that there would be support for Xamarin.Essentials and we could just fix and refine these functions there, but unfortunately I could not wait for this and published this project with the hope that we can  move this to MAUI.Essentials. I even received a little bit of criticism in my direction for starting to do this.

@jfversluis, to you and other developers on the Microsoft team, I have great respect and appreciate Xamarin and the community built around it, please let us not take any hasty decisions. If you can, please discuss this issue with the Microsoft team again.

**Comment by jtorvald** [comment]
@dimonovdd @jfversluis Hi! I read the whole discussion and I must say that I feel Dima's frustration. On the other hand I also share Gerald's opinion that the Community Toolkit would be a very good location for this package to be maintained.

I often notice that 'we developers' start some package because we want to solve our problem and that of other developers. The first year goes super well, but after a while we are super busy, hired by some other company or taking a sabbatical because the work stress is affecting our health. Or for example: working on the next big upgrade (like now with Maui) and stop maintaining the first package. People start to complain, issues piling up, nothing gets done and people need to implement the same functionality themselves. For example the [FirebasePushNotificationPlugin](https://github.com/CrossGeeks/FirebasePushNotificationPlugin) plugin didn't get releases for almost two years.

Another reason: in Xamarin Forms it is recommended to have less packages for performance. If this is still the case in Maui this will also help to have it in MCT.

Since MCT is maintained by the community and employees at Microsoft I guess it will get more love and attention than in a separate repo.

By the way: MCT has 9.4K downloads now, and Xamarin.MediaGallery.Permision.Maui 5.5K (how can it be that it was so much in January but hardly increased?)

Anyway @dimonovdd thanks for this package! I'm using it now. It was just a bit of a surprise that it didn't automatically fixed the photo orientation. I needed to solve that in my project myself. I will see if I can do a PR so that users at least can use the functionality when they migrate from Media.Plugin to MediaGallery.

**Reply by dimonovdd** [reply]
@jtorvald Hi, I started that package because Essentials has not received any support for a long time, and many PRs have not yet been accepted. Although it is  official package from Microsoft. The same can happen with any Microsoft repository.


I will be glad to any PRs in MediaGallery.  Yes, I really don't have much time for Open Source right now (I was promoted and this terrible world situation)

**Reply by jtorvald** [reply]
> @jtorvald Hi, I started that package because Essentials has not received any support for a long time, and many PRs have not yet been accepted. Although it is  official package from Microsoft. The same can happen with any Microsoft repository.

I understand, it's frustrating. Just the community toolkit is also managed outside of Microsoft's management budgets so other community members can also review PR's and decide as a community of things will be picked up. As from how I understand it that is. Is that correct @jfversluis ?

> I will be glad to any PRs in MediaGallery.  

Cool will see when I've the time

> Yes, I really don't have much time for Open Source right now (I was promoted and this terrible world situation)

Yeah it's a mess. Congrats on the promotion. Could be a good reason to move it to MCT maybe? Hope we all can go back to normal soon.

**Comment by vhugogarcia** [comment]
Hello friends,

So, this library and functionality will be really good and it will helps thousands of developers around the world. So, I'm wondering what did you decide for the future of this to support MAUI?

Thanks in advance,
Victor

**Reply by jfversluis** [reply]
No final decision has been made. There are options to use this with .NET MAUI today (MediaGallery) and I am working on a specification to bring the missing functionality to .NET MAUI directly, but that probably won't be implemented until .NET 8.

**Reply by vhugogarcia** [reply]
That is fantastic @jfversluis! Thanks for sharing the next steps. 👍🏻 . Looking forward for it.

**Comment by KSemenenko** [comment]
Hi, I just fix [Xamarin.MediaGallery](https://github.com/dimonovdd/Xamarin.MediaGallery) plugin to make it work works for maui.
So maybe I can add it into toolkit and also add some more functions?

Also current media picker is not good at all.
It is not working with custom selects photos on iOS and vertical photos displayed in horizontal way.

what do you think?

**Reply by KSemenenko** [reply]
@jfversluis if you don’t have any pans or maybe you already have some ideas, so we can start implement them :)

**Reply by VladislavAntonyuk** [reply]
MAUI FilePicker allows picking photos, doesn’t it?

**Reply by dimonovdd** [reply]
Who already uses MAUI? Neither vs4mac nor rider support it so far. 

<img width="595" alt="image" src="https://user-images.githubusercontent.com/59065470/198827670-53550607-837e-49b4-8ae0-900698c534bf.png">

**Reply by VladislavAntonyuk** [reply]
I believe it is clear you need to install workloads. Both VS4Mac and rider support MAUI, but in preview

**Reply by jfversluis** [reply]
I am working on a proposal for .NET MAUI to make things better. Let me try and post that this week. Then we can start implementing it all straight in .NET MAUI without the need for any libraries, as it should be :)

**Reply by KSemenenko** [reply]
> MAUI FilePicker allows picking photos, doesn’t it?

I used Media picker, I need to have images and photos, so why I have to use FilePicker?
Also Media Galery is good. It works wideout extra permissions.
Also I can open several photos, so….

@dimonovdd I will create pr then in your library

**Reply by dimonovdd** [reply]
> I believe it is clear you need to install workloads. Both VS4Mac and rider support MAUI, but in preview

preview ??? 

> @dimonovdd I will create pr then in your library

Thanks. I've been very busy lately. I wanted to update the project several times but  stability of MAUI repelled me every time

**Reply by KSemenenko** [reply]
> > I believe it is clear you need to install workloads. Both VS4Mac and rider support MAUI, but in preview
 
> preview ???

Yes, for preview. You relay need to update workloads.

**Reply by dimonovdd** [reply]
everything is updated. I'm afraid to use the preview, it may affect my work

**Reply by KSemenenko** [reply]
> I am working on a proposal for .NET MAUI to make things better. Let me try and post that this week. Then we can start implementing it all straight in .NET MAUI without the need for any libraries, as it should be :)

also I found, it's will be good to have access to galler a list of images, to make preview. and also we need access to preview of images.

**Comment by CliffAgius** [comment]
Closed as part of base Dotnet MAUI - https://learn.microsoft.com/en-us/dotnet/maui/platform-integration/device-media/picker?tabs=wi
